include/ofi_recvwin.h

```diff
@@ -0,0 +1,152 @@
+/*
+ * Copyright (c) 2018 Amazon.com, Inc. or its affiliates. All rights reserved.
+ *
+ * This software is available to you under a choice of one of two
+ * licenses.  You may choose to be licensed under the terms of the GNU
+ * General Public License (GPL) Version 2, available from the file
+ * COPYING in the main directory of this source tree, or the
+ * BSD license below:
+ *
+ *     Redistribution and use in source and binary forms, with or
+ *     without modification, are permitted provided that the following
+ *     conditions are met:
+ *
+ *      - Redistributions of source code must retain the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer.
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+/*
+ * This utility provides a sliding receive window implementation. It uses a
+ * circular queue to order incoming messages based on the message ID in the
+ * transport protocol.
+ */
+
+#if !defined(FI_RECVWIN_H)
+#define FI_RECVWIN_H
+
+#include "config.h"
+
+#include <assert.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include <unistd.h>
+#include <fcntl.h>
+#include <ofi.h>
+#include <ofi_rbuf.h>
+
+#define OFI_DECL_RECVWIN_BUF(entrytype, name)				\
+OFI_DECLARE_CIRQUE(entrytype, recvwin_cirq);				\
+struct name {								\
+	uint64_t exp_msg_id;						\
+	unsigned int win_size;						\
+	struct recvwin_cirq *pending;					\
+};									\
+DECLARE_FREESTACK(struct name, name ##_fs);				\
```

Comment:
I think the recvwin abstraction should first be defined as a stand alone construct.  Providers can then associate the recvwin/recvq with each peer and allocate it however is best.

Reply:
My thinking was that providers will end up doing something like this anyway during init time and allocate and return them based on the endpoint lifecycle. But fair point, I can leave it up to the consumer to decide how they want to use it.